Removed Venmo `fallbacktoWeb`

stechiu opened the pull request:
Removed Venmo `fallbacktoWeb` opt-in to default to universal link

### Summary of changes

- Removed `fallbacktoWeb` and replaced with universal link

### Checklist

- [x] Added a changelog entry

### Authors
> List GitHub usernames for everyone who contributed to this pull request.

- @stechiu

jaxdesmarais reviewed (commented):
We will want to change the base branch to `v7`. I think we can also clean up the logic in `BTVenmoAppSwitchRedirectURL` to only use universal links.

scannillo commented:
Like Jax mentioned, since this change removes something from the SDKs public interface, it will break a merchant app who might be integrating with this property. So this PR should target `v7` branch

stechiu commented:
> Like Jax mentioned, since this change removes something from the SDKs public interface, it will break a merchant app who might be integrating with this property. So this PR should target `v7` branch

I totally forgot to change the target branch. Just fixed!

stechiu commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -67,7 +67,7 @@ import BraintreeCore
     // swiftlint:disable:next function_body_length cyclomatic_complexity
     public func tokenize(_ request: BTVenmoRequest, completion: @escaping (BTVenmoAccountNonce?, Error?) -> Void) {
-        linkType = request.fallbackToWeb ? .universal : .deeplink
+        linkType = .universal
```
Good point, I was overthinking

stechiu commented:
TODO: Fix failing tests

jaxdesmarais reviewed (commented):
Looks like when we changed the base branch, that branch wasn't up to date with the latest from main. We will need to merge the latest from `main` into `v7` then merge the latest from `v7` into your branch to remove that portion of the diff from this PR.

stechiu commented:
> Looks like when we changed the base branch, that branch wasn't up to date with the latest from main. We will need to merge the latest from `main` into `v7` then merge the latest from `v7` into your branch to remove that portion of the diff from this PR.

Updated. How often do we merge changes from main branch to feature branches?

jaxdesmarais commented:
> Updated. How often do we merge changes from main branch to feature branches?

I normally try to update long lived feature branches once a week. If a large feature is merged in we may want to update it sooner, but that's generally what I've done previously.

warmkesselj commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -428,13 +418,13 @@ import BraintreeCore
             throw BTVenmoError.disabled
         }
 
 
```
Nit: Looks an extra space here

warmkesselj commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -100,7 +99,7 @@ import BraintreeCore
             
             do {
-                _ = try self.verifyAppSwitch(with: configuration, fallbackToWeb: request.fallbackToWeb)
+                _ = try self.verifyAppSwitch(with: configuration)
```
Does verifyAppSwitch need to bother with returning any value?

stechiu replied:
I assumed it would still need to return `true` since `verifyAppSwitch` checks if Venmo is enabled and if Venmo app is installed. Let me know if you think it's ok to remove the Bool return otherwise

warmkesselj replied:
If ```verifyAppSwitch``` is only being used here I think it should be safe. I think the function will just throw an error  if not enabled, and the returned value isn't being used.

warmkesselj commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -43,8 +43,8 @@ import BraintreeCore
-    /// Used for sending the type of flow, universal vs deeplink to FPTI
-    private var linkType: LinkType?
+    /// Used for sending the type of flow, universal, to FPTI
+    private var linkType: LinkType = .universal
```
Is linkType always universal and if so do we still need this top level property?

stechiu replied:
It will always be universal

> do we still need this top level property?
`linkType` is used throughout `BTVenmoClient` so we'll still need the global property for sure. Where else do you suggest to move it?

stechiu replied:
Good call, I've made the changes!

jaxdesmarais reviewed (commented):
Can we update `BTVenmoAppSwitchRedirectURL` to only have references to the universal link method?

stechiu commented:
> Can we update `BTVenmoAppSwitchRedirectURL` to only have references to the universal link method?

Good call, I've removed `urlSchemeURL()`

jaxdesmarais commented on CHANGELOG.md:
```diff
@@ -1,9 +1,13 @@
-## unreleased
+## unreleased (v7)
+* BraintreeVenmo
+    * Remove `fallbacktoWeb` property from the `verifyAppSwitch` method
```
```suggestion
    * Remove `fallbacktoWeb` property from the `verifyAppSwitch` method; all Venmo flows will now use universal links to switch to the Venmo app or fallback to the web flow if the Venmo app is not installed
```

jaxdesmarais commented on Sources/BraintreePayPal/BTPayPalClient.swift:
```diff
@@ -62,9 +62,6 @@ import BraintreeDataCollector
     /// Used for analytics purposes, to determine if brower-presentation event is associated with a locally cached, or remotely fetched `BTConfiguration`
     private var isConfigFromCache: Bool?
 
-    /// Used for sending the type of flow, universal vs deeplink to FPTI
-    private var linkType: LinkType?
-
```
We will still want to keep the link type for the PayPal flow, but we can remove it from Venmo.

stechiu replied:
Oh ok, I'll need to add the LinkType enum back in then

scannillo replied:
LinkType is a little confusing to me. Does it represent the type of the URL that we are opening? Or the URL that returns back to the SDK? Might be worth clarifying in the docstring

stechiu replied:
It's the type of URL link we're opening when a buyer clicks to purchase with __ (PayPal in this case). I'll clarify in the docstring

jaxdesmarais commented on Sources/BraintreePayPal/BTPayPalReturnURL.swift:
```diff
@@ -42,7 +42,7 @@ struct BTPayPalReturnURL {
         url.scheme == "https" && (url.path.contains("cancel") || url.path.contains("success"))
     }
 
-    static func isValidURLAction(url: URL, linkType: LinkType?) -> Bool {
```
Same here, we will want to keep this for the PayPal flow

jaxdesmarais commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -428,13 +413,13 @@ import BraintreeCore
 
 
-        if !fallbackToWeb && !isVenmoAppInstalled() {
+        if !isVenmoAppInstalled() {
```
We can remove this if block entirely, as we have it now we would throw an error if the app is not installed, which we don't want to do since we should always fallback to web now. We can also remove the entire `isVenmoAppInstalled` method in this file.

jaxdesmarais reviewed (commented):
Are we able to verify things are working as expected with these changes? If you have the Venmo app installed it should switch to the app and back and if the app is removed we should fallback to the default browser.

agedd commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -428,13 +413,13 @@ import BraintreeCore
     // MARK: - App Switch Methods
 
-    func verifyAppSwitch(with configuration: BTConfiguration, fallbackToWeb: Bool) throws -> Bool {
+        func verifyAppSwitch(with configuration: BTConfiguration) throws -> Bool {
```
small nit: remove the extra space at the beginning 👀

stechiu commented:
> Are we able to verify things are working as expected with these changes? If you have the Venmo app installed it should switch to the app and back and if the app is removed we should fallback to the default browser.

Yep :) It looks great

https://github.com/user-attachments/assets/a2c056e6-cf52-48a4-9280-69e02c8e42a3

scannillo commented on Braintree.podspec:
```diff
@@ -15,7 +15,7 @@ Pod::Spec.new do |s|
   s.source           = { :git => "https://github.com/braintree/braintree_ios.git", :tag => s.version.to_s }
 
-  s.platform         = :ios, "16.0"
+  s.platform         = :ios, "14.0"
```
This shouldn't change back to 14 (probably a merge error)

scannillo commented on Sources/BraintreeCore/BTHTTP.swift:
```diff
@@ -52,7 +52,7 @@ class BTHTTP: NSObject, URLSessionTaskDelegate {
 
     var acceptLanguageString: String {
-        "\(Locale.current.language.languageCode?.identifier ?? "en")-\(Locale.current.region?.identifier ?? "US")"
+        "\(Locale.current.languageCode ?? "en")-\(Locale.current.regionCode ?? "US")"
```
You might have to rebase, or merge in `v7`. Looks like a lot of old changes weren't in your branch

scannillo commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@
   * Bump minimum supported deployment target to iOS 16+
 * BraintreeVenmo
   * Update `BTVenmoRequest` to make all properties accessible on the initializer only vs via the dot syntax.
+  * Remove `fallbacktoWeb` property from the `verifyAppSwitch` method; all Venmo flows will now use universal links to switch to the Venmo app or fallback to the web flow if the Venmo app is not installed
```
`verifyAppSwitch` is an internal method, so a merchant isn't going to be concerned with it. Instead, it's more important to them that the `fallbackToWeb` setting was removed from the public `BTVenmoRequest` class

stechiu replied:
Thanks! I'll update

warmkesselj reviewed (approved):
Approving ahead of time. Just this final comment on my end: https://github.com/braintree/braintree_ios/pull/1434#discussion_r1819705263

agedd reviewed (approved):
lgtm!

jaxdesmarais commented on Sources/BraintreeVenmo/BTVenmoClient.swift:
```diff
@@ -100,10 +99,9 @@ import BraintreeCore
+                _ = try self.verifyAppSwitch(with: configuration)
             } catch {
                 self.notifyFailure(with: error, completion: completion)
-                return
```
I think we will want to add back this return, my guess is that this is why these tests are failing:
```
	BTVenmoClient_Tests.testTokenize_whenVenmoRequest_setsVaultAnalyticsTag()
	BTVenmoClient_Tests.testTokenizeVenmoAccount_whenVenmoConfigurationDisabled_callsBackWithError()
	BTVenmoClient_Tests.testTokenizeVenmoAccount_whenVenmoConfigurationDisabled_callsBackWithError()
	BTVenmoClient_Tests.testTokenizeVenmoAccount_whenVenmoConfigurationDisabled_callsBackWithError()
	BTVenmoClient_Tests.testTokenizeVenmoAccount_whenVenmoConfigurationMissing_callsBackWithError()
	BTVenmoClient_Tests.testTokenizeVenmoAccount_whenVenmoConfigurationMissing_callsBackWithError()
	BTVenmoClient_Tests.testTokenizeVenmoAccount_whenVenmoConfigurationMissing_callsBackWithError()
```

stechiu replied:
Just added it back and tested locally. Seems like removing `return` was the cause of the failed tests

scannillo commented on Demo/UI Tests/Venmo UI Tests/Venmo_UITests.swift:
```diff
@@ -4,60 +4,16 @@ class Venmo_UITests: XCTestCase {
         demoApp.launch()
     }
-    
-    func testTokenizeVenmo_whenSignInSuccessfulWithPaymentContext_returnsNonce() {
```
We should add a TODO + JIRA ticket for investigating what we want to do about UI tests now that our MockVenmo app will no longer work. (Also likely removing the MockVenmo app entirely, unless we can rig it to mimic Venmo's universal link)

stechiu replied:
Good point, done!

stechiu replied:
Ticket https://paypal.atlassian.net/jira/software/c/projects/DTMOBILES/boards/27751?selectedIssue=DTMOBILES-1237
